js/app.js

```diff
@@ -1256,106 +1101,205 @@ const main = () => {
+              // update lyric position
+              if (!l1Player.status.playing.id) break;
+              const currentSeconds = msg.data.pos;
+              let lastObject = null;
+              let lastObjectTrans = null;
+              $scope.lyricArray.forEach((lyric) => {
+                if (currentSeconds >= lyric.seconds / 1000) {
+                  if (lyric.translationFlag !== true) {
+                    lastObject = lyric;
+                  } else {
+                    lastObjectTrans = lyric;
+                  }
+                }
+              });
+              if (lastObject && lastObject.lineNumber !== $scope.lyricLineNumber) {
+                const lineElement = $(
+                  `.page .playsong-detail .detail-songinfo .lyric p[data-line="${lastObject.lineNumber}"]`,
+                )[0];
+                const windowHeight = $(
+                  '.page .playsong-detail .detail-songinfo .lyric',
+                ).height();
+
+                const adjustOffset = 30;
+                const offset = lineElement.offsetTop - windowHeight / 2 + adjustOffset;
+                $('.lyric').animate(
+                  {
+                    scrollTop: `${offset}px`,
+                  },
+                  500,
+                );
+                $scope.lyricLineNumber = lastObject.lineNumber;
+                if (lastObjectTrans
+                  && lastObjectTrans.lineNumber !== $scope.lyricLineNumberTrans) {
+                  $scope.lyricLineNumberTrans = lastObjectTrans.lineNumber;
+                }
+                if (isElectron()) {
+                  const { ipcRenderer } = require('electron');
+                  const currentLyric = $scope.lyricArray[lastObject.lineNumber].content;
+                  let currentLyricTrans = '';
+                  if ($scope.enableLyricFloatingWindowTranslation === true && lastObjectTrans) {
+                    currentLyricTrans = $scope.lyricArray[lastObjectTrans.lineNumber].content;
+                  }
+                  ipcRenderer.send('currentLyric', { lyric: currentLyric, tlyric: currentLyricTrans });
+                }
+              }
+
+              // 'currentTrack:duration'
+              (() => {
+                const durationSec = Math.floor(msg.data.duration);
+                const durationStr = `${Math.floor(durationSec / 60)}:${(`0${durationSec % 60}`).substr(-2)}`;
+                if (msg.data.duration === 0 || $scope.currentDuration === durationStr) {
+                  return;
+                }
+                $scope.currentDuration = durationStr;
+              })();
+
+              // 'track:progress'
+              if ($scope.changingProgress === false) {
+                $scope.$evalAsync(() => {
+                  if (msg.data.duration === 0) {
+                    $scope.myProgress = 0;
+                  } else {
+                    $scope.myProgress = Math.round(msg.data.pos / msg.data.duration * 100);
```

Comment:
$scope.myProgress = msg.data.pos / msg.data.duration * 100;

Math.round will lose precision after dot

Reply:
Precision issue fixed.
As for the refresh rate, currently is 10 times per second. The background thread will send an event in each tick as well as update the MediaSession state.
On the extension page, the below code can be used to get the current play position directly. But due to the overhead of communication between web pages, if invoking this method too frequently, such as using in `requestAnimationFrame()` callback, will cause an observable performance issue.
`chrome.extension.getBackgroundPage().player.currentHowl.seek()`